File: graph.js
import FastPriorityQueue from "fastpriorityqueue";

class Graph {
    constructor() {
        this.vertices = {};
    }

    addVertex(vertex) {
        if (!(vertex in this.vertices)) {
            this.vertices[vertex] = {};
        }
    }

    addEdge(source, destination, weight) {
        if (!(source in this.vertices)) this.addVertex(source);
        if (!(destination in this.vertices)) this.addVertex(destination);

        this.vertices[source][destination] = weight;
        this.vertices[destination][source] = weight;
    }

    getNeighbors(vertex) {
        if (vertex in this.vertices) {
            return this.vertices[vertex];
        } else {
            return {};
        }
    }

    getShortestDistance(start) {
        const distances = {};
        for (let vertex in this.vertices) {
            distances[vertex] = Infinity;
        }
        distances[start] = 0;

        const pq = new FastPriorityQueue((a, b) => a[0] < b[0]);
        pq.add([0, start]);

        while (pq.size > 0) {
            const [currentDistance, currentVertex] = pq.poll();
            if (currentDistance > distances[currentVertex]) {
                continue;
            }
            for (let neighbor in this.vertices[currentVertex]) {
                const weight = this.vertices[currentVertex][neighbor];
                const distance = currentDistance + weight;

                if (distance < distances[neighbor]) {
                    distances[neighbor] = distance;
                    pq.add([distance, neighbor]);
                }
            }
        }
        return distances;
    }
}




// Example usage:
const graph = new Graph();

graph.addVertex('A');
graph.addVertex('B');
graph.addVertex('C');

graph.addEdge('A', 'B', 5);
graph.addEdge('B', 'C', 3);
graph.addEdge('A', 'C', 10);

console.log(graph.vertices);  // Output: {'A': {'B': 5, 'C': 10}, 'B': {'A': 5, 'C': 3}, 'C': {'B': 3, 'A': 10}}

const distances = graph.getShortestDistance('A');
console.log('Distances from "A":', distances);

File: graph.py
import heapq

class Graph:
    def __init__(self):
        self.vertices = {}

    def add_vertex(self, vertex):
        if vertex not in self.vertices:
            self.vertices[vertex] = {}

    def add_edge(self, source, destination, weight):
        if source in self.vertices and destination in self.vertices:
            self.vertices[source][destination] = weight
            self.vertices[destination][source] = weight  # For undirected graph

    def get_neighbors(self, vertex):
        if vertex in self.vertices:
            return self.vertices[vertex]
        else:
            return {}

    def get_shortest_path(self, start):
        distances = {vertex: float('inf') for vertex in self.vertices}
        distances[start] = 0
        pq = [(0, start)]

        while pq:
            current_distance, current_vertex = heapq.heappop(pq)
            if current_distance > distances[current_vertex]:
                continue
            for neighbor, weight in self.vertices[current_vertex].items():
                distance = current_distance + weight
                if distance < distances[neighbor]:
                    distances[neighbor] = distance
                    heapq.heappush(pq, (distance, neighbor))

        return distances



# Example usage:
graph = Graph()

graph.add_vertex('A')
graph.add_vertex('B')
graph.add_vertex('C')

graph.add_edge('A', 'B', 5)
graph.add_edge('B', 'C', 3)
graph.add_edge('A', 'C', 10)

print(graph.vertices)  # Output: {'A': {'B': 5, 'C': 10}, 'B': {'A': 5, 'C': 3}, 'C': {'B': 3, 'A': 10}}

distances = graph.get_shortest_path('A')
print('Shortest distances from "A":', distances)